rust: add golang bindings

ffmancera opened the pull request:
This patch is introducing golang bindings for the C bindings of
nmstate. The following functions are being introduced:

 - RetrieveNetState(flags int)
   * Returns (state, error) where state is a string containing the
     current state.

 - ApplyNetState(flags int, state string)
   * Returns (state, error) where state is a string containing the
     desired state applied.

This needs to handle the logs better as currently they are being shown
only when Nmstate fails on the operation.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera committed:
rust: add golang bindings

This patch is introducing golang bindings for the C bindings of
nmstate. The following functions are being introduced:

 - RetrieveNetState()
   * Returns (state, error) where state is a string containing the
     current state.

 - ApplyNetState(state string)
   * Returns (state, error) where state is a string containing the
     desired state applied.

 - CommitCheckpoint(checkpoint string)
   * Returns (checkpoint, error) where checkpoint is the string
     containing the checkpoint path committed.

 - RollbackCheckpoint(checkpoint string)
   * Returns (checkpoint, error) where checkpoint is the string
     containing the checkpoint path used for rolling back.

The logs are written to the io.Writer passed by the user. This way the
user can use the logs as they prefer.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera commented:
@qinqon Could you take a look? Thanks!

cathay4t commented on rust/src/go/pkg/nmstate/Makefile:
```diff
@@ -0,0 +1,16 @@
+	go build;
+
+install: build
+	go install;
```
You need to honor `DESTDIR` and maybe other variable for rpmbuild.

qinqon replied:
We don't have a binary to install this is just a golang lib that users will compile before use it, there is no need for an RPM.

ffmancera replied:
I've dropped the install step. Thanks!

cathay4t commented on rust/src/go/pkg/nmstate/nmstate.go:
```diff
@@ -0,0 +1,53 @@
+	}()
+	if rc != 0 {
+		//TODO: Handle the logs properly
+		return "", fmt.Errorf("failed applying nmstate net state %s with rc: %d and logs: %s", state, rc, C.GoString(log))
```
The error kind is not exposed, does go have error type stuff?

ffmancera replied:
According to @qinqon the error kind is not used in golang. So we don't need it, if any customer is willing to use it they can request it later. But the logs needs to be written properly.

cathay4t commented on rust/src/go/pkg/nmstate/nmstate_test.go:
```diff
@@ -0,0 +1,19 @@
+import "github.com/stretchr/testify/assert"
+
+const NMSTATE_FLAG_NONE = 0
+const NMSTATE_FLAG_KERNEL_ONLY = 1 << 1
```
There are 3 constants I forgot to expose in C header:

```
const NMSTATE_FLAG_NO_VERIFY: u32 = 1 << 2;
const NMSTATE_FLAG_INCLUDE_STATUS_DATA: u32 = 1 << 3;
const NMSTATE_FLAG_INCLUDE_SECRETS: u32 = 1 << 4;

```

ffmancera replied:
I exposed all the flags now.

qinqon commented on rust/src/go/pkg/nmstate/go.mod:
```diff
@@ -0,0 +1,5 @@
+module github.com/nmstate/nmstate
```
I am trying to tes to integrate this PR at kubernetes-nmstate but I think the module has to be `module github.com/nmstate/nmstate/rust/src/go/pkg/nmstate/` since it's the place where the go.mod is going to be living.

ffmancera replied:
Done, thanks!

qinqon commented on .gitignore:
```diff
@@ -27,3 +27,4 @@ _kubevirtci
 rust/Cargo.lock
 rust/target
 rust/src/clib/test/nmstate_test
+rust/src/go/pkg/nmstate/*.sum
```
The go.sum cannot be excluded, it's the one that pin the dependencies.

ffmancera replied:
Done, it is not excluded anymore.

qinqon commented on rust/Makefile:
```diff
@@ -11,6 +11,7 @@ CLIB_PKG_CONFIG_FILE=nmstate.pc
 CLIB_PKG_CONFIG_FILE_IN=src/clib/$(CLIB_PKG_CONFIG_FILE).in
 PYTHON_MODULE_NAME=nmstate
 PYTHON_MODULE_SRC=src/python/nmstate
+GO_MODULE_SRC=src/go/pkg/nmstate
```
We don't need the "pkg" part we can go directly with `src/go/nmstate` also having it under rust directory is weird too maybe it has to live next to the rust directory at `nmstate/go/nmstate`

qinqon replied:
Maybe we can setup sone "vanity import" at `nmstate.io` so users will reference it and nmstate can change the path where the dep is living without affecting them.

Placing a html somewhere at nmstate.io should be enough and github pages will serve it.

UPDATE: blog post about doing it with github pages https://alexejk.io/article/github-pages-go-vanity-urls/

qinqon commented on rust/Makefile:
```diff
@@ -76,6 +79,9 @@ install: $(CLI_EXEC_RELEASE)
 	    install -p -v -D -m644 $(PYTHON_MODULE_SRC)/*.py \
 		    $(DESTDIR)$(PYTHON3_SITE_DIR)/$(PYTHON_MODULE_NAME)/; \
 	fi
+	if [ $(SKIP_GO_INSTALL) != 1 ];then \
+	    make install -C $(GO_MODULE_SRC); \
+	fi
```
There is no need to install go src, the user points directly to github.com name.

ffmancera replied:
Done, I have dropped it. Thank you!

qinqon commented on rust/src/go/pkg/nmstate/Makefile:
```diff
@@ -0,0 +1,16 @@
+install: build
+	go install;
+
+check:
+	go test;
```
There is no unit test here, we can check linters though.

ffmancera replied:
Wouldn't this run `rust/src/go/pkg/nmstate/nmstate_test.go`? If not, I will drop it too. Thanks!

qinqon replied:
Now that there are unit tests that should run it, although is better to do `go test ./...` so it runs all the unit tests tha are present event a subdirs.

ffmancera replied:
Changed to `go test ./`, thanks!

qinqon commented on rust/src/go/pkg/nmstate/nmstate.go:
```diff
@@ -0,0 +1,53 @@
+package nmstate
+
+// #cgo CFLAGS: -g -Wall
+// #cgo LDFLAGS: -L/usr/local/lib64 -lnmstate -Wl,-rpath=/usr/local/lib64
```
Normally `/usr/local/lib64` is implicit you don't need to specify it, -lnmstate should be enough.

ffmancera replied:
Done, I have dropped it. Currently it is: `// #cgo LDFLAGS: -lnmstate`. Thanks!

qinqon commented on rust/src/go/pkg/nmstate/nmstate.go:
```diff
@@ -0,0 +1,53 @@
+import "C"
+import "fmt"
+
+func RetrieveNetState(flags uint) (string, error) {
```
We can use bitmask at golang, https://play.golang.org/p/_bTfJBuonzs

```suggestion
type Flags uint

const None Flags = 0

const (
    KernelOnly = 1 << iota
)
func RetrieveNetStateWithFlags(flags Flags) (string, error) {
...
}

func RetrieveNetState() (string, error) {
    return RetrieveNetStateWithFlags(None)
}
```

ffmancera replied:
Done, I have introduced bitmask following the pattern you described. Thank you!

qinqon commented on rust/src/go/pkg/nmstate/nmstate.go:
```diff
@@ -0,0 +1,53 @@
+		C.nmstate_log_free(log)
+	}()
+	if rc != 0 {
+		//TODO: Handle the logs properly
```
My two cents is to add a "verbose" bitmask to print the errors or pass an new `io.Writer` argument to print the logs where the user want to, maybe that's the best option.

ffmancera replied:
Done, the user will need to pass a io.Writer. So this way they can use the logs as they prefer.

qinqon reviewed (changes requested):
What do you think about wrapping all with a struct that will have the flags and a writer to print logs ? 

```golang
type Controller struct {
    Flags Flags
    Logs  io.Writer
}

func (c *Controller) RetrieveNetState() (string, error){...}
func (c *Controller) ApplyNetState(state string) (string, error){...}

```

Users normally want to retrieve and apply the same way (at least our case), so they will use this like

```golang
ncl := nmstate.Controller{Flags: nmstate.KernelOnly, Output: os.Stdout}
state, err := ncl.RetrieveNetState()...
if err != nil {...}
err := ncl.ApplyNetState(desiredState)
if err != nil {...}
```

ffmancera committed:
rust, clib: add nmstate_checkpoint_free()

The checkpoint char array must be set free too. This function is needed
for the golang bindings.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

qinqon commented on rust/src/go/nmstate/Makefile:
```diff
@@ -0,0 +1,13 @@
+# This Makefile is handling the go build, install, test and clean. It is needed
```
Let's move this to go top level  `rust/src/go/Makefile`

ffmancera replied:
We are going to keep using `rust/src/go/nmstate/Makefile` as we discussed. Later we can use the vanity imports to simplify the import. Thanks!

qinqon commented on rust/src/go/nmstate/go.mod:
```diff
@@ -0,0 +1,5 @@
+module github.com/nmstate/nmstate/rust/src/go/nmstate
```
Move to `rust/src/go/go.mod`

qinqon commented on rust/src/go/nmstate/go.sum:
```diff
@@ -0,0 +1,10 @@
+github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8=
```
ditto

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+package nmstate
```
Add the CGO build tag so this is compiled only whent CGO is enabled.

```
// +build cgo
```

ffmancera replied:
Done, thanks!

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+
+type Flags uint
+
+const None Flags = 0
+
+const (
+	KernelOnly Flags = 1 << iota
+	NoVerify
+	IncludeStatusData
+	IncludeSecrets
+	NoCommit
+)
```
You can do 

```golang
const (
	None = 0 << iota
        KernelOnly Flags
	NoVerify
	IncludeStatusData
	IncludeSecrets
	NoCommit
)
```

ffmancera replied:
Done, thanks!

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+		C.nmstate_err_kind_free(err_kind)
+		C.nmstate_log_free(log)
+	}()
+	io.WriteString(logs, C.GoString(log))
```
You have handler `WriteString` error https://pkg.go.dev/io#WriteString

```golang
if err := io.WriteString(logs, C.GoString(log)); err != nil {
    return "", fmt.Errorf("failed wiring logs: %v", err)
}
```

ffmancera replied:
Done, thank you! :-)

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+		C.nmstate_err_kind_free(err_kind)
+		C.nmstate_log_free(log)
+	}()
+	io.WriteString(logs, C.GoString(log))
```
ditto

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+		C.nmstate_err_kind_free(err_kind)
+		C.nmstate_log_free(log)
+	}()
+	io.WriterString(logs, C.GoString(log))
```
ditto

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,111 @@
+		C.nmstate_err_kind_free(err_kind)
+		C.nmstate_log_free(log)
+	}()
+	io.WriterString(logs, C.GoString(log))
```
ditto

qinqon commented on rust/src/go/nmstate/nmstate_test.go:
```diff
@@ -0,0 +1,16 @@
+import "testing"
+import "github.com/stretchr/testify/assert"
+
+func TestRetrieveNetState(t *testing.T) {
+	netState, err := RetrieveNetState(None)
+	assert.NoError(t, err, "must succeed calling retrieve_net_state c binding")
+	assert.NotEmpty(t, netState, "net state should not be empty")
+}
+
+func TestRetrieveNetStateKernelOnly(t *testing.T) {
+	netState, err := RetrieveNetState(KernelOnly)
+	assert.NoError(t, err, "must succeed calling retrieve_net_state c binding")
+	assert.NotEmpty(t, netState, "net state should not be empty")
+}
```
Don't we want to test the rest of them ?

ffmancera replied:
Yes, I will create test for the other functions in a follow up PR. I just wanted to introduce very basic test coverage

qinqon reviewed (changes requested):
Add some go docs to the Exported functions.

ffmancera committed:
clib: fix headers of checkpoint functions

The checkpoint functions are currently being exposed as `void` return
type. This is not correct, they should be exposed as `int` return type.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

filanov commented:
@ffmancera Is it possible to add `Generate configurations` as part of this effort?

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,199 @@
+	Timeout	    uint
+	LogsWritter io.Writer
+	Flags	    byte
+	State	    string
+	Checkpoint  string
```
These two has to be arguments of their functions.

ffmancera replied:
Done!

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,199 @@
+)
+
+type Nmstate struct {
+	Timeout	    uint
+	LogsWritter io.Writer
+	Flags	    byte
```
We don't hvae to expose this since they will be set with the `With` function.

ffmancera replied:
Done!

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,199 @@
+	Checkpoint  string
+}
+
+const (
+	none byte = 0 << iota
+	kernel_only
+	no_verify
+	include_status_data
+	include_secrets
+	no_commit
+)
```
```suggestion
const none = 0
const (
	kernelOnly = 1 << iota
	noVerify
	includeStatusData
	includeSecrets
	noCommit
)
```

ffmancera replied:
I noticed that `none` is not needed. I changed the other flags names to use the correct format.

qinqon commented on rust/src/go/nmstate/nmstate.go:
```diff
@@ -0,0 +1,199 @@
+		return "", fmt.Errorf("failed writting logs: %v", err)
+	}
+	if rc != 0 {
+		return "", fmt.Errorf("failed retrieving nmstate net state with rc: %d", rc)
```
We should add `err_msg`and `err_kind` here right ?

ffmancera replied:
Done, thank you!

qinqon commented on rust/src/go/nmstate/nmstate_test.go:
```diff
@@ -0,0 +1,20 @@
+	netState, err := nms.RetrieveNetState()
+	assert.NoError(t, err, "must succeed calling retrieve_net_state c binding")
+	assert.NotEmpty(t, netState, "net state should not be empty")
+}
```
Don't we want to test apply ?

ffmancera replied:
Yes, but we are going to integrate this with the CI later. Let's merge a simple support and basic coverage. Thanks!

ffmancera commented:
> @ffmancera Is it possible to add `Generate configurations` as part of this effort?

Hi! What do you mind by `Generate configuration`, the `gc`  command? It is not implemented on Rust yet so it is not possible. If done later we can add the Golang binding, that won't be a problem. But adding the gc command here is too much.

filanov commented:
> > @ffmancera Is it possible to add `Generate configurations` as part of this effort?
> 
> Hi! What do you mind by `Generate configuration`, the `gc` command? It is not implemented on Rust yet so it is not possible. If done later we can add the Golang binding, that won't be a problem. But adding the gc command here is too much.

Yes the `gc` command. Great later is fine for me as well. 
Is there an issue  that i can follow? and what is the timeline for that? 
I'm using in a another project and it will really help us to have it as well.

ffmancera commented:
> > > @ffmancera Is it possible to add `Generate configurations` as part of this effort?
> > 
> > 
> > Hi! What do you mind by `Generate configuration`, the `gc` command? It is not implemented on Rust yet so it is not possible. If done later we can add the Golang binding, that won't be a problem. But adding the gc command here is too much.
> 
> Yes the `gc` command. Great later is fine for me as well. Is there an issue that i can follow? and what is the timeline for that? I'm using in a another project and it will really help us to have it as well.

I am sorry but there is not a timeline for this. We expect to do it in the following months but we cannot promise that. We are working on providing all the features in nmstate-rust. Binding will implement the features ASAP. Thank you!
